### DISCUSSION TITLE
关于ObservableCollection的使用的问题

### DISCUSSION THREAD

**Discussion by minChaojie** [BYJRK/DotNet-Discussions#26] [MVVM]
在MVVM中我viewModel中有一个ObservableCollection<Student>的属性，现在我改变了Student中的一个属性内容后，信息是无法刷新到界面上的。目前已知的解决办法，第一种是让Student实现INotifyPropertyChanged，但是Model实现INotifyPropertyChanged感觉怪怪的，第二种就是直接把要变更的student信息删除后重新添加。
您这边还有别的什么方法吗？谢谢

**Comment by BYJRK** [comment]
`INotifyPropertyChanged` 这个接口与这个类是 ViewModel 与否并没有什么关系，甚至于和 MVVM 都没有必然的关系，所以不代表实现了接口就意味着成为了 ViewModel 之类。CommunityToolkit 中给 ViewModel 的基类命名为 `ObservableObject`，可见它只表示其中的属性变化可以被观测。那么如果你希望 Model 可以被观测，就可以实现这个接口。

以及，如果你希望 Collection 中的属性被观测到时也可以通知给 Collection，你可以使用 `BindingList`，它会自动注册子类的 `PropertyChanged` 事件。

**Reply by minChaojie** [reply]
嗯嗯好的谢谢

**Reply by love8gege** [reply]
是不是 之前使用 
 [ObservableProperty]   
    private ObservableCollection<T> oc;
改为 
[ObservableProperty]   
    private BindingList<T> bl; 
这样来实现？

**Reply by BYJRK** [reply]
用法上来说没什么区别，以及如果你想要观测集合中子项的变化，你还需要注册 `BindingList` 的 `ListChanged` 事件，然后可以在事件中去通知相应的属性。

另外，除非你的这个 Collection 本身有可能发生变化，否则其实没有必要把它也弄成 `ObservableProperty`

**Reply by mingjun0315** [reply]
大佬，在新的一期视频里面使用BindingList确实实现了Employee子项Salary的实时更新，但是这是因为Salary属性已经在ObservableCollection的测试时已经改造为可通知的状态。我这边测试了一下，如果Salary只是get\set，那么好像也不能实现对子项的更新。

**Reply by mingjun0315** [reply]
我

> 用法上来说没什么区别，以及如果你想要观测集合中子项的变化，你还需要注册 `BindingList` 的 `ListChanged` 事件，然后可以在事件中去通知相应的属性。
> 
> 另外，除非你的这个 Collection 本身有可能发生变化，否则其实没有必要把它也弄成 `ObservableProperty`

请问我应该怎么写，才可以注册 BindingList 的 ListChanged 事件，然后可以在事件中去通知相应的属性呢？

**Reply by BYJRK** [reply]
> 大佬，在新的一期视频里面使用BindingList确实实现了Employee子项Salary的实时更新，但是这是因为Salary属性已经在ObservableCollection的测试时已经改造为可通知的状态。我这边测试了一下，如果Salary只是get\set，那么好像也不能实现对子项的更新。

我当时做视频的时候脑子似乎不太清醒……Salary发生改变确实是因为它的setter里面进行了Notify，而不是因为使用了BindingList（因为前面用ObservableCollection的时候也实现了相同的效果）。我主要是想讲，BindingList可以主动注册子项的PropertyChanged事件，并且触发自己的ListChanged事件

**Reply by BYJRK** [reply]
> 我
> 
> > 用法上来说没什么区别，以及如果你想要观测集合中子项的变化，你还需要注册 `BindingList` 的 `ListChanged` 事件，然后可以在事件中去通知相应的属性。
> > 另外，除非你的这个 Collection 本身有可能发生变化，否则其实没有必要把它也弄成 `ObservableProperty`
> 
> 请问我应该怎么写，才可以注册 BindingList 的 ListChanged 事件，然后可以在事件中去通知相应的属性呢？

我觉得你的问题可能就有点类似我视频里的修改 Salary 通知 TotalSalary 吧？如果是的话，那就是同样的实现方式即可，在 ViewModel 的构造里面注册事件
